packages/block-library/src/navigation/index.php

```diff
@@ -212,16 +212,28 @@ function block_core_navigation_build_html( $attributes, $block, $colors, $font_s
 			$css_classes .= ' ' . $class_name;
 		};
 
+		$item_tag = 'div';
+		if ( isset( $block['attrs']['id'] ) || isset( $block['attrs']['url'] ) ) {
+			$item_tag = 'a';
+		}
+
 		$html .= '<li class="' . esc_attr( $css_classes . ( $has_submenu ? ' has-child' : '' ) ) .
 			( $is_active ? ' current-menu-item' : '' ) . '"' . $style_attribute . '>' .
-			'<a class="wp-block-navigation-link__content"';
+			'<' . $item_tag . ' class="wp-block-navigation-link__content"';
 
-		// Start appending HTML attributes to anchor tag.
-		if ( isset( $block['attrs']['url'] ) ) {
-			$html .= ' href="' . esc_url( $block['attrs']['url'] ) . '"';
+		if (
+			isset( $block['attrs']['id'] )
+		) {
+			$html .= ' href="' .
+				esc_url( get_permalink( $block['attrs']['id'] ) ) .
+			'"';
+		} else {
+			if ( 'a' === $item_tag ) {
+				$html .= ' href="' . esc_url( $block['attrs']['url'] ) . '"';
+			}
 		}
 
-		if ( isset( $block['attrs']['opensInNewTab'] ) && true === $block['attrs']['opensInNewTab'] ) {
+		if ( 'a' === $item_tag && isset( $block['attrs']['opensInNewTab'] ) && true === $block['attrs']['opensInNewTab'] ) {
```

Comment:
Not a showstopper, but I feel like some of this logic could be made more readable. For example extracting the logic for getting the url to a function and then checking for the existence of a url first would tidy up some of the resulting if statement logic:
```php
$url = get_navigation_link_url( $block['attrs'] );
$has_url = ! is_empty( $url );
$item_tag = $has_url ? 'a' : 'div';

// ...

if ( $has_url ) {
    $html .= ' href="' . $url . '"';
}

// ...
if ( $has_url && isset( $block['attrs']['opensInNewTab'] ) && true === $block['attrs']['opensInNewTab'] ) {
```

Reply:
Yep, I've cleaned the code a bit as suggested, thanks!